features/lorawan/LoRaWANStack.cpp

```diff
@@ -0,0 +1,2046 @@
+}
+
+void LoRaWANStack::set_lora_event_cb(mbed::Callback<void(lora_events_t)> event_cb)
+{
+    _events = event_cb;
+}
+
+lora_mac_status_t LoRaWANStack::add_channels(const lora_channelplan_t &channel_plan)
+{
+    // If device is not initialized, stop right away
+    if (_device_current_state == DEVICE_STATE_NOT_INITIALIZED)
+    {
+        tr_error("Stack not initialized!");
+        return LORA_MAC_STATUS_NOT_INITIALIZED;
+    }
+
+    ChannelParams_t mac_layer_ch_params;
+    LoRaMacStatus_t status;
+
+    GetPhyParams_t get_phy;
+    PhyParam_t phy_param;
+    uint8_t max_num_channels;
+
+    // Check first how many channels the selected PHY layer supports
+    get_phy.Attribute = PHY_MAX_NB_CHANNELS;
+    phy_param = lora_phy.get_phy_params(&get_phy);
+    max_num_channels = (uint8_t) phy_param.Value;
+
+    // check if user is setting more channels than supported
+    if (channel_plan.nb_channels > max_num_channels) {
+        return LORA_MAC_STATUS_PARAMETER_INVALID;
+    }
+
+    for (uint8_t i = 0; i < channel_plan.nb_channels; i++) {
+        mac_layer_ch_params.Band = channel_plan.channels[i].ch_param.band;
+        mac_layer_ch_params.DrRange.Fields.Max = channel_plan.channels[i].ch_param.dr_range.lora_mac_fields_s.max;
+        mac_layer_ch_params.DrRange.Fields.Min = channel_plan.channels[i].ch_param.dr_range.lora_mac_fields_s.min;
+        mac_layer_ch_params.DrRange.Value = channel_plan.channels[i].ch_param.dr_range.value;
+        mac_layer_ch_params.Frequency = channel_plan.channels[i].ch_param.frequency;
+        mac_layer_ch_params.Rx1Frequency =channel_plan.channels[i].ch_param.rx1_frequency;
+
+        status = LoRaMacChannelAdd(channel_plan.channels[i].id, mac_layer_ch_params);
+
+        if (status != LORAMAC_STATUS_OK) {
+            return error_type_converter(status);
+        }
+    }
+
+    return LORA_MAC_STATUS_OK;
+}
+
+lora_mac_status_t LoRaWANStack::drop_channel_list()
+{
+    if (_device_current_state == DEVICE_STATE_NOT_INITIALIZED )
+    {
+        tr_error("Stack not initialized!");
+        return LORA_MAC_STATUS_NOT_INITIALIZED;
+    }
+
+    lora_mac_status_t status = LORA_MAC_STATUS_OK;
+
+    GetPhyParams_t get_phy;
+    PhyParam_t phy_param;
+    uint8_t max_num_channels;
+    uint16_t *channel_masks;
+    uint16_t *default_channel_masks;
+
+    // Check first how many channels the selected PHY layer supports
+    get_phy.Attribute = PHY_MAX_NB_CHANNELS;
+    phy_param = lora_phy.get_phy_params(&get_phy);
+    max_num_channels = (uint8_t) phy_param.Value;
+
+    // Now check the channel mask for enabled channels
+    get_phy.Attribute = PHY_CHANNELS_MASK;
+    phy_param = lora_phy.get_phy_params(&get_phy);
+    channel_masks = phy_param.ChannelsMask;
+
+    // Now check the channel mask for default channels
+    get_phy.Attribute = PHY_CHANNELS_DEFAULT_MASK;
+    phy_param = lora_phy.get_phy_params(&get_phy);
+    default_channel_masks = phy_param.ChannelsMask;
```

Comment:
Since channel_masks/default_channel_masks are pointers it would be better to check if they are NULL before using it below.